kedro/pipeline/node.py

```diff
@@ -681,3 +683,27 @@ def _get_readable_func_name(func: Callable) -> str:
         name = "<partial>"
 
     return name
+
+
+def _unpack_params(
+    func: Callable, inputs: None | str | list[str] | dict[str, str]
+) -> None | str | list[str] | dict[str, str]:
+    """Iterate over Node inputs to see if they need to be unpacked.
+
+    Returns:
+        Either original inputs if no input was unpacked or a list of inputs if an input was unpacked.
+    """
+    use_new = False
+    new_inputs = []
+    for idx, _input in enumerate(_to_list(inputs)):
+        if _input.startswith("**"):
+            use_new = True
+            dict_root = _input.split(":")[-1]
+            _func_arguments = [arg for arg in inspect.signature(func).parameters]
```

Comment:
So there is a soft assumption that the dictionary keys match the function arguments. Do we want to throw a warning or error if:

(a) No arguments match
(b) Partial arguments match

Reply:
So that is an assumption, I suppose we could also implement tuple unpacking if a user wanted to use index instead of key-value pairs.

The current behavior of no match/partial match is consistent I think with what users would expect. If you pass the dictionary `model_options` and the function expects a key `features` and that key is missing you would get the following error before the run.

`ValueError: Pipeline input(s) {'params:model_options.features'} not found in the DataCatalog`

Reply:
true! interested to see if others have an opinion though

Reply:
I would vote for adding at least basic validation because there are definitely cases in which this method will not work correctly, for example, when having a `**kwargs'—only node function. Then, we could also extend tests with negative samples so it's clear which cases we cover and which we do not.

Reply:
I'm a little confused about how we would implement any validation at this point in the process, this function is called when the Node object is being initialized and I don't believe the catalog is available to verify against is it?


> there are definitely cases in which this method will not work correctly, for example, when having a `**kwargs`—only node function.

Can you help me understand this?  if your function was something like `def foo(float_1, integer_2, string_3):` you could pass `**params:dictionary` and it would work. If your function was `def foo(**kwargs)` then this implementation wouldn't work but I wasn't under the impression that was what was being asked for, because you could just change it to `def foo(kwargs)` and then access the values normally from there.

Reply:
Sure. I don't mean that this is the expected usage example and we aim to make it work. But it doesn't mean that someone cannot make that (or something else that breaks the current unpacking method) by mistake if syntax allows. There might be some other cases as well. To simplify the validation, we can follow the approach proposed above - check no/partial argument match and add some possible incorrect test cases to ensure the validation catches them. At the same time extend the docs with the expected usage example.

Reply:
I think this is the main outstanding discussion right?